File: reference-python/backend/planners/rrt.py
"""RRT — rastgele örneklemeli yol arama."""

import random
import time
from typing import Optional, Sequence

from backend.planners.base import Obstacle, PlanResult, Planner, Point


def _distance(a: Point, b: Point) -> float:
    return sum((x - y) ** 2 for x, y in zip(a, b)) ** 0.5


class RRTPlanner(Planner):
    """Rastgele ağaç genişletme (Rapidly-exploring Random Tree) ile yol arar."""

    name = "rrt"

    def __init__(
        self,
        collision_checker,
        step_size: float = 0.1,
        max_iterations: int = 3000,
        goal_bias: float = 0.1,
        goal_tolerance: float = 0.05,
        workspace_margin: float = 0.3,
        segment_resolution: float = 0.02,
        rng: Optional[random.Random] = None,
    ) -> None:
        super().__init__(collision_checker)
        self.step_size = step_size
        self.max_iterations = max_iterations
        self.goal_bias = goal_bias
        self.goal_tolerance = goal_tolerance
        self.workspace_margin = workspace_margin
        self.segment_resolution = segment_resolution
        self._rng = rng or random.Random()

    def _workspace_bounds(self, start: Point, goal: Point):
        return [
            (min(s, g) - self.workspace_margin, max(s, g) + self.workspace_margin)
            for s, g in zip(start, goal)
        ]

    def _sample(self, bounds) -> Point:
        return tuple(self._rng.uniform(low, high) for low, high in bounds)

    def _nearest(self, nodes: Sequence[Point], point: Point) -> Point:
        return min(nodes, key=lambda node: _distance(node, point))

    def _steer(self, origin: Point, target: Point) -> Point:
        dist = _distance(origin, target)
        if dist <= self.step_size:
            return target
        ratio = self.step_size / dist
        return tuple(o + (t - o) * ratio for o, t in zip(origin, target))

    def _segment_free(self, a: Point, b: Point) -> bool:
        dist = _distance(a, b)
        steps = max(1, int(dist / self.segment_resolution))
        for step in range(1, steps + 1):
            t = step / steps
            point = tuple(a[i] + (b[i] - a[i]) * t for i in range(3))
            if not self.collision_checker(point):
                return False
        return True

    def _build_path(self, parent: dict, end_node: Point, start: Point, goal: Point) -> list:
        path_nodes = []
        current = end_node
        while current is not None:
            path_nodes.append(current)
            current = parent[current]
        path_nodes.reverse()
        if end_node != goal:
            path_nodes.append(goal)
        return path_nodes

    def plan(self, start: Point, goal: Point, obstacles: Sequence[Obstacle]) -> PlanResult:
        started_at = time.perf_counter()

        if not self.collision_checker(start) or not self.collision_checker(goal):
            return PlanResult(
                success=False, elapsed_seconds=time.perf_counter() - started_at, algorithm=self.name
            )

        bounds = self._workspace_bounds(start, goal)
        nodes = [start]
        parent = {start: None}
        nodes_expanded = 0

        for _ in range(self.max_iterations):
            nodes_expanded += 1
            sample = goal if self._rng.random() < self.goal_bias else self._sample(bounds)
            nearest = self._nearest(nodes, sample)
            new_point = self._steer(nearest, sample)

            if new_point in parent:
                continue
            if not self.collision_checker(new_point) or not self._segment_free(nearest, new_point):
                continue

            nodes.append(new_point)
            parent[new_point] = nearest

            if _distance(new_point, goal) <= self.goal_tolerance:
                path = self._build_path(parent, new_point, start, goal)
                return PlanResult(
                    success=True,
                    path=path,
                    elapsed_seconds=time.perf_counter() - started_at,
                    nodes_expanded=nodes_expanded,
                    algorithm=self.name,
                )

        return PlanResult(
            success=False,
            path=[],
            elapsed_seconds=time.perf_counter() - started_at,
            nodes_expanded=nodes_expanded,
            algorithm=self.name,
        )

File: lib/robotics/planners/rrt.ts
import type { Vec3 } from "../transform";
import type { CollisionChecker, Planner, PlanResult } from "./base";
import { distance } from "./base";

export interface RrtOptions {
  /** Her adımda ağacın en fazla ne kadar uzayacağı (metre). */
  stepSize?: number;
  maxIterations?: number;
  /** Rastgele örnek yerine doğrudan hedefin seçilme olasılığı. */
  goalBias?: number;
  goalTolerance?: number;
  /** Örnekleme kutusunun start/goal kutusunun dışına taşma payı. */
  workspaceMargin?: number;
  segmentResolution?: number;
  /**
   * Düzlemsel planlama: z ekseni start.z değerinde sabitlenir.
   *
   * Neden gerekli: sahneler (PlannerRace) üstten görünen 2B bir çalışma
   * alanı çiziyor, ama planlayıcı 3B örnekliyordu. Engeller z ekseninde
   * sınırlı olduğu için yol, kullanıcının göremediği üçüncü boyuttan
   * dolaşıp engeli "delmiş" gibi görünüyordu. 2B sahnelerde bu açık
   * kapatılmalı.
   */
  planar?: boolean;
  /** [0, 1) aralığında sözde-rastgele sayı üreten fonksiyon; testlerde tohumlanabilir. */
  random?: () => number;
}

type Bounds = readonly [number, number][];

/**
 * RRT — rastgele ağaç genişletme (Rapidly-exploring Random Tree).
 * reference-python/backend/planners/rrt.py'nin TS portu (bkz. docs/02).
 *
 * Not: RNG dizisi Python'ın `random.Random` çıktısıyla bit bit eşleşmez
 * (farklı algoritma), bu yüzden bu planlayıcı Python fixture'ına karşı değil,
 * özellik testlerine (yol geçerliliği, çarpışmasızlık) karşı doğrulanır —
 * bkz. docs/02-mimari.md "Doğrulama stratejisi" katman 3.
 */
export class RrtPlanner implements Planner {
  name = "rrt";
  protected stepSize: number;
  protected maxIterations: number;
  protected goalBias: number;
  protected goalTolerance: number;
  protected workspaceMargin: number;
  protected segmentResolution: number;
  protected planar: boolean;
  protected random: () => number;

  constructor(options: RrtOptions = {}) {
    this.stepSize = options.stepSize ?? 0.1;
    this.maxIterations = options.maxIterations ?? 3000;
    this.goalBias = options.goalBias ?? 0.1;
    this.goalTolerance = options.goalTolerance ?? 0.05;
    this.workspaceMargin = options.workspaceMargin ?? 0.3;
    this.segmentResolution = options.segmentResolution ?? 0.02;
    this.planar = options.planar ?? false;
    this.random = options.random ?? Math.random;
  }

  protected workspaceBounds(start: Vec3, goal: Vec3): Bounds {
    const axes: (keyof Vec3)[] = ["x", "y", "z"];
    return axes.map((axis) => {
      // Düzlemsel modda z hiç örneklenmez: aralık start.z'de kapatılır.
      if (this.planar && axis === "z") return [start.z, start.z] as const;
      const low = Math.min(start[axis], goal[axis]) - this.workspaceMargin;
      const high = Math.max(start[axis], goal[axis]) + this.workspaceMargin;
      return [low, high] as const;
    }) as Bounds;
  }

  protected sample(bounds: Bounds): Vec3 {
    return {
      x: bounds[0][0] + this.random() * (bounds[0][1] - bounds[0][0]),
      y: bounds[1][0] + this.random() * (bounds[1][1] - bounds[1][0]),
      z: bounds[2][0] + this.random() * (bounds[2][1] - bounds[2][0]),
    };
  }

  protected nearest(nodes: readonly Vec3[], point: Vec3): Vec3 {
    let best = nodes[0];
    let bestDist = distance(best, point);
    for (let i = 1; i < nodes.length; i++) {
      const d = distance(nodes[i], point);
      if (d < bestDist) {
        best = nodes[i];
        bestDist = d;
      }
    }
    return best;
  }

  protected steer(origin: Vec3, target: Vec3): Vec3 {
    const dist = distance(origin, target);
    if (dist <= this.stepSize) return target;
    const ratio = this.stepSize / dist;
    return {
      x: origin.x + (target.x - origin.x) * ratio,
      y: origin.y + (target.y - origin.y) * ratio,
      z: origin.z + (target.z - origin.z) * ratio,
    };
  }

  protected segmentFree(a: Vec3, b: Vec3, isFree: CollisionChecker): boolean {
    const dist = distance(a, b);
    // ceil — bkz. collision.ts isSegmentFree: asagi yuvarlama ince engelleri atlıyordu.
    const steps = Math.max(1, Math.ceil(dist / this.segmentResolution));
    for (let step = 1; step <= steps; step++) {
      const t = step / steps;
      const point: Vec3 = { x: a.x + (b.x - a.x) * t, y: a.y + (b.y - a.y) * t, z: a.z + (b.z - a.z) * t };
      if (!isFree(point)) return false;
    }
    return true;
  }

  /** parent zincirini end'den köke doğru izler; end goal değilse sona goal'i ekler. */
  protected buildPath(parent: Map<Vec3, Vec3 | null>, endNode: Vec3, goal: Vec3): Vec3[] {
    const nodes: Vec3[] = [];
    let current: Vec3 | null = endNode;
    while (current !== null) {
      nodes.push(current);
      current = parent.get(current) ?? null;
    }
    nodes.reverse();
    if (endNode !== goal) nodes.push(goal);
    return nodes;
  }

  plan(start: Vec3, goal: Vec3, isFree: CollisionChecker): PlanResult {
    const startedAt = Date.now();
    if (!isFree(start) || !isFree(goal)) {
      return { success: false, path: [], elapsedMs: Date.now() - startedAt, nodesExpanded: 0, algorithm: this.name };
    }

    const bounds = this.workspaceBounds(start, goal);
    const nodes: Vec3[] = [start];
    const parent = new Map<Vec3, Vec3 | null>([[start, null]]);
    let nodesExpanded = 0;

    for (let i = 0; i < this.maxIterations; i++) {
      nodesExpanded++;
      const sample = this.random() < this.goalBias ? goal : this.sample(bounds);
      const nearest = this.nearest(nodes, sample);
      const newPoint = this.steer(nearest, sample);

      if (parent.has(newPoint)) continue;
      if (!isFree(newPoint) || !this.segmentFree(nearest, newPoint, isFree)) continue;

      nodes.push(newPoint);
      parent.set(newPoint, nearest);

      // Hedefe yeterince yaklaşmak tek başına yetmez: son sıçrama
      // (newPoint → goal) da çarpışmasız olmalı. Aksi hâlde goalTolerance
      // kadar uzunlukta, hiç kontrol edilmemiş bir segment yola ekleniyor
      // ve engelin içinden geçen bir yol "başarılı" dönebiliyordu.
      if (distance(newPoint, goal) <= this.goalTolerance) {
        if (newPoint === goal || this.segmentFree(newPoint, goal, isFree)) {
          const path = this.buildPath(parent, newPoint, goal);
          return { success: true, path, elapsedMs: Date.now() - startedAt, nodesExpanded, algorithm: this.name };
        }
      }
    }

    return { success: false, path: [], elapsedMs: Date.now() - startedAt, nodesExpanded, algorithm: this.name };
  }
}
